README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

<a name="top"></a>

# `lancer`: Linearity Assessment <img src="man/figures/README-logo.png" alt="Hex logo of R package lancer. Logo is a picture of a lancer under a saturated curve exploring the mountains." align="right" width="25%" height="25%" /> <br>and Curve Explorer

<!-- badges: start -->

[![R-Command-check](https://github.com/SLINGhub/lancer/workflows/R-CMD-check/badge.svg)](https://github.com/SLINGhub/lancer/actions)
[![Codecov test
coverage](https://codecov.io/gh/SLINGhub/lancer/branch/main/graph/badge.svg?token=RNlP8VlaL2)](https://codecov.io/gh/SLINGhub/lancer)
[![License:
MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://github.com/SLINGhub/lancer/blob/main/LICENSE.md)
[![Maintainer:
JauntyJJS](https://img.shields.io/badge/Maintainer-JauntyJJS-blue.svg)](https://github.com/JauntyJJS)
[![](https://img.shields.io/badge/doi-10.6084/m9.figshare.21966896-blue.svg?style=flat&labelColor=whitesmoke&logo=data%3Aimage%2Fpng%3Bbase64%2CiVBORw0KGgoAAAANSUhEUgAAAB8AAAAfCAYAAAAfrhY5AAAJsklEQVR42qWXd1DTaRrHf%2BiB2Hdt5zhrAUKz4IKEYu9IGiGFFJJQ0gkJCAKiWFDWBRdFhCQUF3UVdeVcRQEBxUI3yY9iEnQHb3bdW1fPubnyz%2F11M7lvEHfOQee2ZOYzPyDv%2B3yf9%2Fk95YX4fx%2BltfUt08GcFEuPR4U9hDDZ%2FVngIlhb%2FSiI6InkTgLzgDcgfvtnovhH4BzoVlrbwr55QnhCtBW4QHXnFrZbPBaQoBh4%2FSYH2EnpBEtqcDMVzB93wA%2F8AFwa23XFGcc8CkT3mxz%2BfXWtq9T9IQlLIXYEuHojudb%2BCM7Hgdq8ydi%2FAHiBXyY%2BLjwFlAEnS6Jnar%2FvnQVhvdzasad0eKvWZKe8hvDB2ofLZ%2FZEcWsh%2BhyIuyO5Bxs2iZIE4nRv7NWAb0EO8AC%2FWPxjYAWuOEX2MSXZVgPxzmRL3xKz3ScGpx6p6QnOx4mDIFqO0w6Q4fEhO5IzwxlSwyD2FYHzwAW%2BAZ4fEsf74gCumykwNHskLM7taQxLYjjIyy8MUtraGhTWdkfhkFJqtvuVl%2F9l2ZquDfEyrH8B0W06nnpH3JtIyRGpH1iJ6SfxDIHjRXHJmdQjLpfHeN54gnfFx4W9QRnovx%2FN20aXZeTD2J84hn3%2BqoF2Tqr14VqTPUCIcP%2B5%2Fly4qC%2BUL3sYxSvNj1NwsVYPsWdMUfomsdkYm3Tj0nbV0N1wRKwFe1MgKACDIBdMAhPE%2FwicwNWxll8Ag40w%2BFfhibJkGHmutjYeQ8gVlaN%2BjO51nDysa9TwNUFMqaGbKdRJZFfOJSp6mkRKsv0rRIpEVWjAvyFkxNOEpwvcAVPfEe%2Bl8ojeNTx3nXLBcWRrYGxSRjDEk0VlpxYrbe1ZmaQ5xuT0u3r%2B2qe5j0J5uytiZPGsRL2Jm32AldpxPUNJ3jmmsN4x62z1cXrbedXBQf2yvIFCeZrtyicZZG2U2nrrBJzYorI2EXLrvTfCSB43s41PKEvbZDEfQby6L4JTj%2FfIwam%2B4%2BwucBu%2BDgNK05Nle1rSt9HvR%2FKPC4U6LTfvUIaip1mjIa8fPzykii23h2eanT57zQ7fsyYH5QjywwlooAUcAdOh5QumgTHx6aAO7%2FL52eaQNEShrxfhL6albEDmfhGflrsT4tps8gTHNOJbeDeBlt0WJWDHSgxs6cW6lQqyg1FpD5ZVDfhn1HYFF1y4Eiaqa18pQf3zzYMBhcanlBjYfgWNayAf%2FASOgklu8bmgD7hADrk4cRlOL7NSOewEcbqSmaivT33QuFdHXj5sdvjlN5yMDrAECmdgDWG2L8P%2BAKLs9ZLZ7dJda%2BB4Xl84t7QvnKfvpXJv9obz2KgK8dXyqISyV0sXGZ0U47hOA%2FAiigbEMECJxC9aoKp86re5O5prxOlHkcksutSQJzxZRlPZmrOKhsQBF5zEZKybUC0vVjG8PqOnhOq46qyDTDnj5gZBriWCk4DvXrudQnXQmnXblebhAC2cCB6zIbM4PYgGl0elPSgIf3iFEA21aLdHYLHUQuVkpgi02SxFdrG862Y8ymYGMvXDzUmiX8DS5vKZyZlGmsSgQqfLub5RyLNS4zfDiZc9Edzh%2FtCE%2BX8j9k%2FqWB071rcZyMImne1SLkL4GRw4UPHMV3jjwEYpPG5uW5fAEot0aTSJnsGAwHJi2nvF1Y5OIqWziVCQd5NT7t6Q8guOSpgS%2Fa1dSRn8JGGaCD3BPXDyQRG4Bqhu8XrgAp0yy8DMSvvyVXDgJcJTcr1wQ2BvFKf65jqhvmxXUuDpGBlRvV36XvGjQzLi8KAKT2lYOnmxQPGorURSV0NhyTIuIyqOmKTMhQ%2BieEsgOgpc4KBbfDM4B3SIgFljvfHF6cef7qpyLBXAiQcXvg5l3Iunp%2FWv4dH6qFziO%2BL9PbrimQ9RY6MQphEfGUpOmma7KkGzuS8sPUFnCtIYcKCaI9EXo4HlQLgGrBjbiK5EqMj2AKWt9QWcIFMtnVvQVDQV9lXJJqdPVtUQpbh6gCI2Ov1nvZts7yYdsnvRgxiWFOtNJcOMVLn1vgptVi6qrNiFOfEjHCDB3J%2BHDLqUB77YgQGwX%2Fb1eYna3hGKdlqJKIyiE4nSbV8VFgxmxR4b5mVkkeUhMgs5YTi4ja2XZ009xJRHdkfwMi%2BfocaancuO7h%2FMlcLOa0V%2FSw6Dq47CumRQAKhgbOP8t%2BMTjuxjJGhXCY6XpmDDFqWlVYbQ1aDJ5Cptdw4oLbf3Ck%2BdWkVP0LpH7s9XLPXI%2FQX8ws%2Bj2In63IcRvOOo%2BTTjiN%2BlssfRsanW%2B3REVKoavBOAPTXABW4AL7e4NygHdpAKBscmlDh9Jysp4wxbnUNna3L3xBvyE1jyrGIkUHaqQMuxhHElV6oj1picvgL1QEuS5PyZTEaivqh5vUCKJqOuIgPFGESns8kyFk7%2FDxyima3cYxi%2FYOQCj%2F%2B9Ms2Ll%2Bhn4FmKnl7JkGXQGDKDAz9rUGL1TIlBpuJr9Be2JjK6qPzyDg495UxXYF7JY1qKimw9jWjF0iV6DRIqE%2B%2FeWG0J2ofmZTk0mLYVd4GLiFCOoKR0Cg727tWq981InYynvCuKW43aXgEjofVbxIqrm0VL76zlH3gQzWP3R3Bv9oXxclrlO7VVtgBRpSP4hMFWJ8BrUSBCJXC07l40X4jWuvtc42ofNCxtlX2JH6bdeojXgTh5TxOBKEyY5wvBE%2BACh8BtOPNPkApjoxi5h%2B%2FFMQQNpWvZaMH7MKFu5Ax8HoCQdmGkJrtnOiLHwD3uS5y8%2F2xTSDrE%2F4PT1yqtt6vGe8ldMBVMEPd6KwqiYECHDlfbvzphcWP%2BJiZuL5swoWQYlS%2Br7Yu5mNUiGD2retxBi9fl6RDGn4Ti9B1oyYy%2BMP5G87D%2FCpRlvdnuy0PY6RC8BzTA40NXqckQ9TaOUDywkYsudxJzPgyDoAWn%2BB6nEFbaVxxC6UXjJiuDkW9TWq7uRBOJocky9iMfUhGpv%2FdQuVVIuGjYqACbXf8aa%2BPeYNIHZsM7l4s5gAQuUAzRUoT51hnH3EWofXf2vkD5HJJ33vwE%2FaEWp36GHr6GpMaH4AAPuqM5eabH%2FhfG9zcCz4nN6cPinuAw6IHwtvyB%2FdO1toZciBaPh25U0ducR2PI3Zl7mokyLWKkSnEDOg1x5fCsJE9EKhH7HwFNhWMGMS7%2BqxyYsbHHRUDUH4I%2FAheQY7wujJNnFUH4KdCju83riuQeHU9WEqNzjsJFuF%2FdTDAZ%2FK7%2F1WaAU%2BAWymT59pVMT4g2AxcwNa0XEBDdBDpAPvgDIH73R25teeuAF5ime2Ul0OUIiG4GpSAEJeYW9wDTf43wfwHgHLKJoPznkwAAAABJRU5ErkJggg%3D%3D)](https://doi.org/10.6084/m9.figshare.21966896)

<!-- badges: end -->

R package used to validate if a curve is linear or has signal
suppression by statistical analysis and plots.

## Table of Content

- [Installation](#installation)
- [Meta](#meta)
- [Overview](#overview)
- [Motivation](#motivation)
- [How It Works](#how-it-works)
- [Workflow Proposed](#workflow-proposed)
- [Usage](#usage)
- [Output Results](#output-results)

## Installation

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("SLINGhub/lancer")
```

If you want to use a proper release version, referenced by a Git tag
(example: 0.1.1) install the package as follows:

``` r
devtools::install_github("SLINGhub/lancer", ref = "0.1.1")
```

<a href="#top">Back to top</a>

## Meta

- We welcome contributions from general questions to bug reports. Check
  out the [contributions](CONTRIBUTING.md) guidelines. Please note that
  this project is released with a [Contributor Code of
  Conduct](https://www.contributor-covenant.org/version/2/0/code_of_conduct/).
  By participating in this project you agree to abide by its terms.
- License:
  [MIT](https://github.com/SLINGhub/lancer/blob/main/LICENSE.md)
- Think `lancer` is useful? Let others discover it, by telling them in
  person, via Twitter
  [![Tweet](https://img.shields.io/twitter/url/http/shields.io.svg?style=social)](https://twitter.com/LOGIN)
  or a blog post.
- Refer to the [NEWS.md
  file](https://github.com/SLINGhub/lancer/blob/main/NEWS.md) to see
  what is being worked on as well as update to changes between back to
  back versions.
- Images used for hex logos are taken from the following sources:
  - [Black Knight by
    J4p4n](https://www.1001freedownloads.com/free-clipart/black-knight)
    in [1001FreeDownloads](https://www.1001freedownloads.com/) released
    under CC0. [![License:
    CC0-1.0](https://licensebuttons.net/l/zero/1.0/80x15.png)](http://creativecommons.org/publicdomain/zero/1.0/)
  - [Mountain
    Range](https://publicdomainvectors.org/en/free-clipart/Mountain-range-clip-art/88586.html)
    in [publicdomainvectors.org](https://publicdomainvectors.org/)
    released under CC0. [![License:
    CC0-1.0](https://licensebuttons.net/l/zero/1.0/80x15.png)](http://creativecommons.org/publicdomain/zero/1.0/)

<a href="#top">Back to top</a>

## Overview

Using curve data as a running example, an overview of the workflow is as
follows

<img src="man/figures/README-Workflow.png" alt="Overview of lancer workflow. Input Curve Signal Data and Curve Annotation are first merged to a Curve Table. Next, Curve Statistics Summary are then calculated. Finally, results are exported as Excel, pdf file or html folder." width="100%" />

<a href="#top">Back to top</a>

## Motivation

The Pearson correlation coefficient has been used widely to test for
linearity. However, it is insufficient as indicated in [Francisco Raposo
(2016)](https://www.sciencedirect.com/science/article/abs/pii/S0165993615301242)

Consider a curve in the linear, saturation and noise regime.

``` r
linear_data <- data.frame(
  conc_var = c(
    10, 25, 40, 50, 60,
    75, 80, 100, 125, 150
  ),
  signal_var = c(
    25463, 63387, 90624, 131274, 138069,
    205353, 202407, 260205, 292257, 367924
  )
)

saturation_regime_data <- data.frame(
  conc_var = c(
    10, 25, 40, 50, 60,
    75, 80, 100, 125, 150
  ),
  signal_var = c(

NEWS.md
# lancer 0.1.1.9000 (development version)

## TODO

# lancer 0.2.0




R/cognostics.R
#' @title Validate Cognostics Data
#' @description Validate Cognostics Data
#' @param cog_df A data frame or tibble that contains cognostics information
#' @param needed_column A vector consisting of needed column names that
#' must be found in `cog_df`,
#' Default:
#' `c("col_name_vec", "desc_vec", "type_vec")`
#' @return An error if the things in `needed_column`
#' is not found in the Cognostics Data
#' @examples
#' # Create Cognostics Dataframe
#' col_name_vec <- c("Curve_Name", "Curve_Batch_Name")
#'
#' desc_vec <- c("Curve_Name", "Curve_Batch_Name")
#'
#' type_vec <- c("factor", "factor")
#'
#' cog_df <- data.frame(
#'   col_name_vec = col_name_vec,
#'   desc_vec = desc_vec,
#'   type_vec = type_vec
#' )
#'
#' validate_cog_df(cog_df)
#' @rdname validate_cog_df
#' @export
validate_cog_df <- function(cog_df,
                            needed_column = c(
                              "col_name_vec",
                              "desc_vec",
                              "type_vec"
                            )) {


  # Check if things in needed_column are in cog_df
  assertable::assert_colnames(cog_df, needed_column,
    only_colnames = FALSE, quiet = TRUE
  )
}


#' @title Create Default Cognostics Data Frame
#' @description Create default cognostics data frame to be used to
#' convert columns in `curve_summary` to class cognostics
#' @return A dataframe of default cognostics information
#' @details Internal function to create a dataframe of default cognostics
#' information to be used by the `trelliscopejs::cog`.
#' @examples
#' cog_df <- create_default_cog_df()
#' cog_df
#' @rdname create_default_cog_df
#' @export
create_default_cog_df <- function() {

  # More details in `trelliscopejs::cog` can be found in
  # <https://rdrr.io/cran/trelliscopejs/man/cog.html>

  col_name_vec <- c(
    "Curve_Name", "Curve_Batch_Name",
    "Curve_Class",
    "wf1_group", "wf2_group",
    "r_corr", "pra_linear", "mandel_p_val",
    "r2_linear", "r2_adj_linear", "mandel_stats",
    "adl_value"
  )

  desc_vec <- c(
    "Curve_Name", "Curve_Batch_Name",
    "Classes of Curves",
    "Group from workflow 1", "Group from workflow 2",
    "Pearson Correlation R values",
    "Linear Regression Percent Residual Accuracy",
    "P values for Mandel test",
    "Linear Regression R^2 Value",
    "Linear Regression Adjusted R^2 Value",
    "Test statistics for Mandel Test",
    "Average Deviation from Linearity"
  )

  type_vec <- c(
    "factor", "factor",
    "factor", "factor", "factor",
    "numeric", "numeric", "numeric",
    "numeric", "numeric", "numeric", "numeric"
  )

  cog_df <- data.frame(
    col_name_vec = col_name_vec,
    desc_vec = desc_vec,
    type_vec = type_vec
  )

  return(cog_df)
}

#' @title Update Cognostics Manually
#' @description Update cognostics on `curve_summary` based on
#' the cognostics parameters given by `cog_df`. We assume
#' `cog_df` is created manually
#' @param curve_summary The summary data frame or tibble generated
#' by function [summarise_curve_table()] and/or
#' [evaluate_linearity()].
#' @param cog_df A data frame or tibble that contains cognostics information
#' If no input is given the cognostics information generated by function
#' [create_default_cog_df()] will be used.
#' Default: NULL
#' @param col_name_vec Column name in `cog_df` to indicate the columns
#' in `curve_summary` that needs to be converted to a cognostics.
#' Default: 'col_name_vec'
#' @param desc_vec Column name in `cog_df` to indicate the description
#' for each cognostics as define in `trelliscopejs::cog`.
#' Default: 'desc_vec'
#' @param type_vec Column name in `cog_df` to indicate the type
#' of each cognostics as define in `trelliscopejs::cog`.
#' Default: 'type_vec'
#' @return `curve_summary` with some columns converted
#' to type cog as defined in `trelliscopejs::cog`.
#' @details
#' More details in `trelliscopejs::cog` can be found in
#' <https://rdrr.io/cran/trelliscopejs/man/cog.html>.
#' @examples
#'
#' # Data Creation
#' concentration <- c(
#'   10, 20, 25, 40, 50, 60,
#'   75, 80, 100, 125, 150,
#'   10, 25, 40, 50, 60,
#'   75, 80, 100, 125, 150
#' )
#'
#' curve_batch_name <- c(
#'   "B1", "B1", "B1", "B1", "B1",
#'   "B1", "B1", "B1", "B1", "B1", "B1",
#'   "B2", "B2", "B2", "B2", "B2",
#'   "B2", "B2", "B2", "B2", "B2"
#' )
#'
#' sample_name <- c(
#'   "Sample_010a", "Sample_020a",
#'   "Sample_025a", "Sample_040a", "Sample_050a",
#'   "Sample_060a", "Sample_075a", "Sample_080a",
#'   "Sample_100a", "Sample_125a", "Sample_150a",
#'   "Sample_010b", "Sample_025b",
#'   "Sample_040b", "Sample_050b", "Sample_060b",
#'   "Sample_075b", "Sample_080b", "Sample_100b",
#'   "Sample_125b", "Sample_150b"
#' )
#'
#' curve_1_saturation_regime <- c(
#'   5748124, 16616414, 21702718, 36191617,
#'   49324541, 55618266, 66947588, 74964771,
#'   75438063, 91770737, 94692060,
#'   5192648, 16594991, 32507833, 46499896,
#'   55388856, 62505210, 62778078, 72158161,
#'   78044338, 86158414
#' )
#'
#' curve_2_good_linearity <- c(
#'   31538, 53709, 69990, 101977, 146436, 180960,
#'   232881, 283780, 298289, 344519, 430432,
#'   25463, 63387, 90624, 131274, 138069,
#'   205353, 202407, 260205, 292257, 367924
#' )
#'
#' curve_3_noise_regime <- c(
#'   544, 397, 829, 1437, 1808, 2231,
#'   3343, 2915, 5268, 8031, 11045,
#'   500, 903, 1267, 2031, 2100,
#'   3563, 4500, 5300, 8500, 10430
#' )
#'
#' curve_4_poor_linearity <- c(
#'   380519, 485372, 478770, 474467, 531640, 576301,
#'   501068, 550201, 515110, 499543, 474745,
#'   197417, 322846, 478398, 423174, 418577,
#'   426089, 413292, 450190, 415309, 457618
#' )
#'
#' curve_batch_annot <- tibble::tibble(
#'   Sample_Name = sample_name,
#'   Curve_Batch_Name = curve_batch_name,
#'   Concentration = concentration
#' )
#'
#' curve_data <- tibble::tibble(
#'   Sample_Name = sample_name,
#'   `Curve_1` = curve_1_saturation_regime,
#'   `Curve_2` = curve_2_good_linearity,
#'   `Curve_3` = curve_3_noise_regime,
#'   `Curve_4` = curve_4_poor_linearity
#' )
#'
#' # Create curve table
#' curve_table <- create_curve_table(
#'   curve_batch_annot = curve_batch_annot,
#'   curve_data_wide = curve_data,
#'   common_column = "Sample_Name",
#'   signal_var = "Signal",
#'   column_group = "Curve_Name"
#' )
#'
#' # Create curve statistical summary
#' curve_summary <- curve_table |>
#'   summarise_curve_table(
#'     grouping_variable = c(
#'       "Curve_Name",
#'       "Curve_Batch_Name"
#'     ),
#'     conc_var = "Concentration",
#'     signal_var = "Signal"
#'   ) |>
#'   dplyr::arrange(.data[["Curve_Name"]]) |>
#'   evaluate_linearity(grouping_variable = c(
#'     "Curve_Name",
#'     "Curve_Batch_Name"
#'   ))
#'
#' # Create our own cog_df
#' col_name_vec <- c(
#'   "Curve_Name", "Curve_Batch_Name",
#'   "Curve_Class"
#' )
#'
#' desc_vec <- c(
#'   "Curve_Name", "Curve_Batch_Name",
#'   "Curve_Class"
#' )
#'
#' type_vec <- c("factor", "factor", "factor")
#'
#' cog_df <- data.frame(
#'   col_name_vec = col_name_vec,
#'   desc_vec = desc_vec,
#'   type_vec = type_vec
#' )
#'
#' updated_summary <- update_cog_manual(
#'   curve_summary,
#'   cog_df,
#'   col_name_vec = "col_name_vec",
#'   desc_vec = "desc_vec",
#'   type_vec = "type_vec"
#' )
#'
#' # Observe that the first two columns has been converted
#' # to class cognostics
#' curve_summary
#'
#' updated_summary
#'
#' @rdname update_cog_manual
#' @export
update_cog_manual <- function(curve_summary,
                              cog_df = NULL,
                              col_name_vec = "col_name_vec",
                              desc_vec = "desc_vec",
                              type_vec = "type_vec") {

  if (is.null(cog_df)) {
    return(curve_summary)
  }

  # Check if cog_df is valid with the relevant columns
  validate_cog_df(cog_df,
    needed_column = c(
      col_name_vec,
      desc_vec,
      type_vec
    )
  )

  # See this webpage to learn how to mutate specific columns.
  # https://stackoverflow.com/questions/52998471/
  # dynamically-determine-if-a-dataframe-column-exists-
  # and-mutate-if-it-does

  for (colname in colnames(curve_summary)) {
    row_index <- which(cog_df$col_name_vec == colname)

    if (length(row_index) == 1) {
      curve_summary <- curve_summary |>
        dplyr::mutate(
          dplyr::across(
            dplyr::one_of(cog_df[[col_name_vec]][row_index]),
            ~ trelliscopejs::cog(.x,
              desc = cog_df[[desc_vec]][row_index],
              type = cog_df[[type_vec]][row_index]
            )
          )
        )
    }
  }
  return(curve_summary)
}

#' Update Cognostics Automatically
#' @description Update cognostics on the `curve_summary` based on
#' the cognostics data frame given by the function `create_default_cog_df`.
#' @param curve_summary The summary table generated
#' by function [summarise_curve_table()] and/or
#' [evaluate_linearity()].
#' @return `curve_summary` with some columns converted
#' to type cog as defined in `trelliscopejs::cog`.
#' @details
#' More details in `trelliscopejs::cog` can be found in
#' <https://rdrr.io/cran/trelliscopejs/man/cog.html>.
#' @examples
#'
#' # Data Creation
#' concentration <- c(
#'   10, 20, 25, 40, 50, 60,
#'   75, 80, 100, 125, 150,
#'   10, 25, 40, 50, 60,
#'   75, 80, 100, 125, 150
#' )
#'
#' curve_batch_name <- c(
#'   "B1", "B1", "B1", "B1", "B1",
#'   "B1", "B1", "B1", "B1", "B1", "B1",
#'   "B2", "B2", "B2", "B2", "B2",
#'   "B2", "B2", "B2", "B2", "B2"
#' )
#'
#' sample_name <- c(
#'   "Sample_010a", "Sample_020a",
#'   "Sample_025a", "Sample_040a", "Sample_050a",
#'   "Sample_060a", "Sample_075a", "Sample_080a",
#'   "Sample_100a", "Sample_125a", "Sample_150a",
#'   "Sample_010b", "Sample_025b",
#'   "Sample_040b", "Sample_050b", "Sample_060b",
#'   "Sample_075b", "Sample_080b", "Sample_100b",
#'   "Sample_125b", "Sample_150b"
#' )
#'
#' curve_1_saturation_regime <- c(
#'   5748124, 16616414, 21702718, 36191617,
#'   49324541, 55618266, 66947588, 74964771,
#'   75438063, 91770737, 94692060,
#'   5192648, 16594991, 32507833, 46499896,
#'   55388856, 62505210, 62778078, 72158161,
#'   78044338, 86158414
#' )
#'
#' curve_2_good_linearity <- c(
#'   31538, 53709, 69990, 101977, 146436, 180960,
#'   232881, 283780, 298289, 344519, 430432,
#'   25463, 63387, 90624, 131274, 138069,
#'   205353, 202407, 260205, 292257, 367924
#' )
#'
#' curve_3_noise_regime <- c(
#'   544, 397, 829, 1437, 1808, 2231,
#'   3343, 2915, 5268, 8031, 11045,
#'   500, 903, 1267, 2031, 2100,
#'   3563, 4500, 5300, 8500, 10430
#' )
#'
#' curve_4_poor_linearity <- c(
#'   380519, 485372, 478770, 474467, 531640, 576301,
#'   501068, 550201, 515110, 499543, 474745,
#'   197417, 322846, 478398, 423174, 418577,
#'   426089, 413292, 450190, 415309, 457618
#' )
#'
#' curve_batch_annot <- tibble::tibble(
#'   Sample_Name = sample_name,
#'   Curve_Batch_Name = curve_batch_name,
#'   Concentration = concentration
#' )
#'
#' curve_data <- tibble::tibble(
#'   Sample_Name = sample_name,
#'   `Curve_1` = curve_1_saturation_regime,
#'   `Curve_2` = curve_2_good_linearity,
#'   `Curve_3` = curve_3_noise_regime,
#'   `Curve_4` = curve_4_poor_linearity
#' )
#'
#' # Create curve table
#' curve_table <- create_curve_table(
#'   curve_batch_annot = curve_batch_annot,
#'   curve_data_wide = curve_data,
#'   common_column = "Sample_Name",
#'   signal_var = "Signal",
#'   column_group = "Curve_Name"
#' )
#'
#' # Create curve statistical summary
#' curve_summary <- curve_table |>
#'   summarise_curve_table(
#'     grouping_variable = c(
#'       "Curve_Name",
#'       "Curve_Batch_Name"
#'     ),
#'     conc_var = "Concentration",
#'     signal_var = "Signal"
#'   ) |>
#'   dplyr::arrange(.data[["Curve_Name"]]) |>
#'   evaluate_linearity(grouping_variable = c(
#'     "Curve_Name",
#'     "Curve_Batch_Name"
#'   ))
#'
#' updated_summary <- update_cog_auto(curve_summary)
#'
#' # Observe that the columns has been converted
#' # to class cognostics
#' curve_summary
#'
#' updated_summary
#'
#' @rdname update_cog_auto
#' @export
update_cog_auto <- function(curve_summary) {

  # Create the default cognostics table
  cog_df <- create_default_cog_df()

  # Update the curve_summary columns and
  # convert the relevant to cognostics
  curve_summary <- curve_summary |>
    update_cog_manual(cog_df = cog_df)

  return(curve_summary)
}

#' @title Convert To Cognostics
#' @description Convert columns in `curve_summary` to
#' `trelliscopejs` cognostics for the `Trelliscope` display
#' @param curve_summary The summary data frame or tibble generated
#' by function [summarise_curve_table()] and/or
#' [evaluate_linearity()]..
#' @param cog_df A data frame or tibble that contains cognostics information
#' If no input is given the cognostics information generated by function
#' [create_default_cog_df()] will be used.
#' Default: NULL
#' @param grouping_variable A character vector of
#' column names in `curve_summary`to indicate how each curve
#' should be grouped by. It is also going to be used as a conditional
#' cognostics in the `trelliscopejs` report.
#' Default: c("Curve_Name", "Curve_Batch_Name")
#' @param panel_variable A column name in `curve_summary` to be converted
#' into a panel for the `Trelliscope` display.
#' Default: NULL
#' @param col_name_vec Column name in `cog_df` to indicate the columns
#' in `curve_summary` that needs to be converted to a cognostics.
#' Default: 'col_name_vec'
#' @param desc_vec Column name in `cog_df` to indicate the description
#' for each cognostics as define in `trelliscopejs::cog`.
#' Default: 'desc_vec'
#' @param type_vec Column name in `cog_df` to indicate the type
#' of each cognostics as define in `trelliscopejs::cog`.
#' Default: 'type_vec'
#' @return The curve summary table with `grouping variable` columns
#' converted to conditional cognostics.
#' other columns in `curve_summary` converted to cognostics
#' to be used in the in the `trelliscopejs` report.
#' @examples
#'
#' # Data Creation
#' concentration <- c(
#'   10, 20, 25, 40, 50, 60,
#'   75, 80, 100, 125, 150,
#'   10, 25, 40, 50, 60,
#'   75, 80, 100, 125, 150
#' )
#'
#' curve_batch_name <- c(
#'   "B1", "B1", "B1", "B1", "B1",
#'   "B1", "B1", "B1", "B1", "B1", "B1",
#'   "B2", "B2", "B2", "B2", "B2",
#'   "B2", "B2", "B2", "B2", "B2"
#' )
#'
#' sample_name <- c(
#'   "Sample_010a", "Sample_020a",
#'   "Sample_025a", "Sample_040a", "Sample_050a",
#'   "Sample_060a", "Sample_075a", "Sample_080a",
#'   "Sample_100a", "Sample_125a", "Sample_150a",
#'   "Sample_010b", "Sample_025b",
#'   "Sample_040b", "Sample_050b", "Sample_060b",
#'   "Sample_075b", "Sample_080b", "Sample_100b",
#'   "Sample_125b", "Sample_150b"
#' )
#'
#' curve_1_saturation_regime <- c(
#'   5748124, 16616414, 21702718, 36191617,
#'   49324541, 55618266, 66947588, 74964771,
#'   75438063, 91770737, 94692060,
#'   5192648, 16594991, 32507833, 46499896,
#'   55388856, 62505210, 62778078, 72158161,
#'   78044338, 86158414
#' )
#'
#' curve_2_good_linearity <- c(
#'   31538, 53709, 69990, 101977, 146436, 180960,
#'   232881, 283780, 298289, 344519, 430432,
#'   25463, 63387, 90624, 131274, 138069,
#'   205353, 202407, 260205, 292257, 367924
#' )
#'
#' curve_3_noise_regime <- c(
#'   544, 397, 829, 1437, 1808, 2231,
#'   3343, 2915, 5268, 8031, 11045,
#'   500, 903, 1267, 2031, 2100,
#'   3563, 4500, 5300, 8500, 10430
#' )
#'
#' curve_4_poor_linearity <- c(
#'   380519, 485372, 478770, 474467, 531640, 576301,
#'   501068, 550201, 515110, 499543, 474745,
#'   197417, 322846, 478398, 423174, 418577,
#'   426089, 413292, 450190, 415309, 457618
#' )
#'
#' curve_batch_annot <- tibble::tibble(
#'   Sample_Name = sample_name,
#'   Curve_Batch_Name = curve_batch_name,
#'   Concentration = concentration
#' )
#'
#' curve_data <- tibble::tibble(
#'   Sample_Name = sample_name,
#'   `Curve_1` = curve_1_saturation_regime,
#'   `Curve_2` = curve_2_good_linearity,
#'   `Curve_3` = curve_3_noise_regime,
#'   `Curve_4` = curve_4_poor_linearity
#' )
#'
#' # Create curve table
#' curve_table <- create_curve_table(
#'   curve_batch_annot = curve_batch_annot,
#'   curve_data_wide = curve_data,
#'   common_column = "Sample_Name",
#'   signal_var = "Signal",
#'   column_group = "Curve_Name"
#' )
#'
#' # Create curve statistical summary
#' curve_summary <- curve_table |>
#'   summarise_curve_table(
#'     grouping_variable = c(
#'       "Curve_Name",
#'       "Curve_Batch_Name"
#'     ),
#'     conc_var = "Concentration",
#'     signal_var = "Signal"
#'   ) |>
#'   dplyr::arrange(.data[["Curve_Name"]]) |>
#'   evaluate_linearity(grouping_variable = c(
#'     "Curve_Name",
#'     "Curve_Batch_Name"
#'   ))
#'
#' updated_summary <- convert_to_cog(
#'   curve_summary,
#'   grouping_variable = c("Curve_Name",
#'                         "Curve_Batch_Name")
#'   )
#'
#' # Observe that the columns has been converted
#' # to class cognostics
#' curve_summary
#'
#' updated_summary
#'
#' @rdname convert_to_cog
#' @export
convert_to_cog <- function(curve_summary,
                           cog_df = NULL,
                           grouping_variable = c(
                             "Curve_Name",
                             "Curve_Batch_Name"
                           ),
                           panel_variable = NULL,
                           col_name_vec = "col_name_vec",
                           desc_vec = "desc_vec",
                           type_vec = "type_vec") {

  # Check if things in needed_column are in curve_summary
  assertable::assert_colnames(curve_summary, grouping_variable,
    only_colnames = FALSE, quiet = TRUE
  )

  # Check if panel_variable is also a grouping variable
  if (isTRUE(panel_variable %in% grouping_variable)) {
    stop(paste(
      "panel_variable", panel_variable,
      "cannot be a grouping_variable"
    ))
  }

  # Convert logical columns to characters
  curve_summary <- curve_summary |>
    dplyr::mutate_if(
      is.logical,
      ~ as.character(.x)
    )

  # Separate the panel variables if it is in curve_summary
  panel_df <- NULL
  if (!is.null(panel_variable)) {
    panel_df <- curve_summary |>
      dplyr::select(dplyr::any_of(c(
        grouping_variable,
        panel_variable
      )))

    curve_summary <- curve_summary |>
      dplyr::select(-dplyr::any_of(c(panel_variable)))
  }

  # Get cognostics for trellis report
  # First convert the columns based on default cognostics
  # from create_default_cog_df()
  # Next convert the columns based on user's input cognostics
  # Lastly, convert the rest of the columns based on class
  # Grouping variables must be the conditional columns
  curve_summary <- curve_summary |>
    update_cog_auto() |>
    update_cog_manual(
      cog_df = cog_df,
      col_name_vec = col_name_vec,
      desc_vec = desc_vec,
      type_vec = type_vec
    ) |>
    trelliscopejs::as_cognostics(
      cond_cols = grouping_variable,
      needs_cond = TRUE, needs_key = FALSE
    )


  # Convert panel variables if any to trelliscope_panels
  if (!is.null(panel_df) &&
    ncol(panel_df) != length(grouping_variable)) {

    # Ensure that the grouping variable is converted to
    # conditional columns
    # Ensure that the panel variable is converted to
    # trelliscope_panel
    panel_df <- panel_df |>
      dplyr::select(dplyr::all_of(c(grouping_variable))) |>
      trelliscopejs::as_cognostics(
        cond_cols = grouping_variable,
        needs_cond = TRUE,
        needs_key = FALSE
      ) |>
      dplyr::bind_cols(panel_df |>
        dplyr::select(dplyr::any_of(c(panel_variable)))) |>
      dplyr::mutate(dplyr::across(
        dplyr::all_of(panel_variable),
        ~ structure(.x, class = c("trelliscope_panels", "list"))
      ))

    # Panel_df to do a left join with curve_summary
    # Move panel_variable to the end
    curve_summary <- panel_df |>
      dplyr::left_join(curve_summary, by = grouping_variable) |>
      dplyr::relocate(dplyr::any_of(c(panel_variable)),
        .after = dplyr::last_col()
      )
  }

  return(curve_summary)
}
